pkg/crd/cloudprivateipconfig/v1/types.go

```diff
@@ -0,0 +1,62 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +genclient:noStatus
+// +resource:path=cloudprivateipconfig
+// +kubebuilder:resource:shortName=cpip,scope=Cluster
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:printcolumn:name="Node Request",type=string,JSONPath=".spec.items[*].node"
+// +kubebuilder:printcolumn:name="IP Request",type=string,JSONPath=".spec.items[*].ip"
+// +kubebuilder:printcolumn:name="Node Assignment",type=string,JSONPath=".status.items[*].node"
+// +kubebuilder:printcolumn:name="IP Assignment",type=string,JSONPath=".status.items[*].ip"
+// CloudPrivateIPConfig is a CRD allowing the user to assign private
+// IP addresses to the primary NIC on cloud VMs. This is done by
+// specifying the spec with the Kubernetes node(s) the request
+// is meant for, as well as the IP(s) requested for that/those
+// node(s).
+type CloudPrivateIPConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the desired private IP request(s).
+	Spec CloudPrivateIPConfigSpec `json:"spec"`
+	// Observed status of CloudPrivateIPConfig. Read-only.
+	// +optional
+	Status CloudPrivateIPConfigStatus `json:"status,omitempty"`
+}
+
+type CloudPrivateIPConfigSpec struct {
+	// The list of per node requested private IPs and their corresponding node assignment.
+	Items []CloudPrivateIPConfigItem `json:"items"`
+}
+
+type CloudPrivateIPConfigStatus struct {
+	// The list of per node status of private IP assignments
+	// executed. In case an assignment fails: the item will be omitted from the
+	// status.items. I.e: if all went well the `.spec` should
+	// equal `.status`
+	Items []CloudPrivateIPConfigItem `json:"items"`
+}
+
+type CloudPrivateIPConfigItem struct {
+	// Node name
+	Node string `json:"node"`
+	// IP address - can be IPv4 or IPv6
+	IP string `json:"ip"`
```

Comment:
maybe add a kubebuilder validation for the IPs?

Reply:
I tried to do this for the egr crd, and failed because its hard to accommodate ipv6 validater, although I'd recommend at least adding some basic validation for both nodename and ip to check its not empty strings.

Reply:
Yes, IPv6 is the main reason I chose not to do so. The validation of all fields is/will be done in the controller.

Reply:
there are some places I think in openshift/api/network/v1 where we may have some precedent for v6 validator. If it's too much hassle, we can leave it. but i liken the kubebuilder validations to compile time checks vs. run time. Better to fail fast.

Reply:
also, does it make sense to split this into a v4 field and a v6 field to make things clear?

Reply:
I've had a look and didn't find any IPv6 validators, maybe because not much IPv6 work has been done for openshift-sdn? Do you know of one such place? 

> also, does it make sense to split this into a v4 field and a v6 field to make things clear?

Nothing else in the Kubernetes eco-sphere makes that distinction, I'm thinking about `pod.spec.podIP / pod.spec.podIPs / service.spec.ClusterIP / service.spec.ClusterIPs / node.spec.nodeIP`, so I don't really think it's needed here. This component as we've already mentioned: should be even less user facing than any of those other fields, so I see an even lower need for it to be done here.